File: C/Algorithms/Sorting/Heap_Sort.c
/*Heap sort implementation in C*/
#include <stdio.h>
#include <stdlib.h>
typedef int TYPE;
int left(int);
int right(int); 
int parent(int);
void max_heapify(TYPE [], int, int);
void build_max_heap(TYPE [], int);
void heap_sort(TYPE [], int);
void print_array(TYPE [], int);

int main(){
	puts("-----Heap Sort-----");
	TYPE a[] = {4, 10, 3, 5, 1};
	int length = sizeof(a) / sizeof(TYPE);
	
	printf("Unsorted: ");
	print_array(a, length);
	heap_sort(a, length);
	
	printf("Sorted: ");
	print_array(a, length);
	
	return EXIT_SUCCESS;
}
void heap_sort(TYPE a[], int length){
	build_max_heap(a, length);
	int i = length - 1;
	do {
		TYPE temp = a[0];
		a[0] = a[i];
		a[i] = temp;
		
		max_heapify(a, 0, --length);
	} while(--i >= 0);
}
void build_max_heap(TYPE a[], int length){
	int i = length/2;
	do {
		max_heapify(a, i, length);
	}while(--i  >= 0); //first i = i - 1 is doen and then i >= 0 afterwards.
}
void max_heapify(TYPE a[], int i, int length){
	int r = right(i);
	int l = left(i);
	int max = i;
	if(l < length && a[max] < a[l]) max = l;
	if(r < length && a[max] < a[r]) max = r;
	if(max != i) {
		TYPE temp = a[i];
		a[i] = a[max];
		a[max] = temp;
		max_heapify(a, max, length);
	}
}
int left(int i) {
	return 2*i + 1;
}
int right(int i) {
	return 2*i + 2;
}
int parent(int i) {
	return (i - 1)/2;
}
void print_array(TYPE a[], int length){
	int i = 0;
	putchar('[');
	while(i < length);
		if(i > 0) printf(", ");
		printf("%d", a[i++]); 
	} 
	puts("]");
}

File: CSharp/Algorithms/Sorting/HeapSort.cs
// C# program for implementation of Heap Sort
using System;

public class HeapSort {
	public void heapSort(int[] arr)
	{
		int n = arr.Length;

		// Build heap (rearrange array)
		for (int i = n / 2 - 1; i >= 0; i--)
			heapify(arr, n, i);

		// One by one extract an element from heap
		for (int i = n - 1; i > 0; i--) {
			// Move current root to end
			int temp = arr[0];
			arr[0] = arr[i];
			arr[i] = temp;

			// call max heapify on the reduced heap
			heapify(arr, i, 0);
		}
	}

	// To heapify a subtree rooted with node i which is
	// an index in arr[]. n is size of heap
	void heapify(int[] arr, int n, int i)
	{
		int largest = i; // Initialize largest as root
		int l = 2 * i + 1; // left = 2*i + 1
		int r = 2 * i + 2; // right = 2*i + 2

		// If left child is larger than root
		if (l < n && arr[l] > arr[largest])
			largest = l;

		// If right child is larger than largest so far
		if (r < n && arr[r] > arr[largest])
			largest = r;

		// If largest is not root
		if (largest != i) {
			int swap = arr[i];
			arr[i] = arr[largest];
			arr[largest] = swap;

			// Recursively heapify the affected sub-tree
			heapify(arr, n, largest);
		}
	}

	/* A utility function to print array of size n */
	static void printArray(int[] arr)
	{
		int n = arr.Length;
		for (int i = 0; i < n; ++i)
			Console.Write(arr[i] + " ");

		Console.Write("\n");
	}

	public static void Main()
	{
		int[] arr = { 66, 31, 14, 55, 10, 7 };
		int n = arr.Length;

		HeapSort ob = new HeapSort();
		ob.heapSort(arr);

		Console.WriteLine("Sorted array is");
		printArray(arr);
	}
}